File: python/modcdp/launcher/BrowserLauncher.py
# MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
# Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
# - ./js/src/launcher/BrowserLauncher.ts
# - ./go/modcdp/launcher/BrowserLauncher.go
from __future__ import annotations

import json
import re
import urllib.request
from collections.abc import Callable
from typing import TYPE_CHECKING, Any, TypeAlias

from ..types.modcdp import LaunchedBrowser, ModCDPLauncherConfig
from ..types.toJSON import modCDPToJSON

if TYPE_CHECKING:
    from ..transport.UpstreamTransport import UpstreamTransport


LauncherConfig: TypeAlias = ModCDPLauncherConfig


DEFAULT_CHROME_READY_TIMEOUT_MS = 45_000
DEFAULT_CHROME_READY_POLL_INTERVAL_MS = 100
CDP_URL_SCHEME_RE = re.compile(r"^[a-z][a-z\d+\-.]*://", re.I)


class BrowserLauncher:
    launched: LaunchedBrowser | None

    def __init__(self, config: LauncherConfig | dict[str, Any] | None = None) -> None:
        self.config = _launcher_config(config)
        self.launched = None

    def update(self, config: LauncherConfig | dict[str, Any] | None = None) -> "BrowserLauncher":
        incoming = _launcher_config(config)
        updates = incoming.model_dump(exclude_unset=True)
        if "launcher_local_args" in incoming.model_fields_set:
            updates["launcher_local_args"] = merge_chrome_args(self.config.launcher_local_args, incoming.launcher_local_args)
        if "launcher_local_extra_args" in incoming.model_fields_set:
            updates["launcher_local_extra_args"] = merge_chrome_args(self.config.launcher_local_extra_args, incoming.launcher_local_extra_args)
        self.config = LauncherConfig.model_validate({**self.config.model_dump(), **updates})
        return self

    def configForUpstream(self) -> dict[str, Any]:
        config: dict[str, Any] = {}
        upstream_ws_cdp_url = (self.launched.cdp_url if self.launched is not None else None) or self.config.launcher_remote_cdp_url
        if upstream_ws_cdp_url:
            config["upstream_ws_cdp_url"] = upstream_ws_cdp_url
        return config

    def configForServer(self, upstream: UpstreamTransport) -> dict[str, Any]:
        launcher_local_loopback_cdp_url = self.launched.loopback_cdp_url if self.launched is not None else None
        if not launcher_local_loopback_cdp_url and upstream.config.upstream_mode == "ws" and upstream.config.upstream_ws_cdp_url:
            launcher_local_loopback_cdp_url = upstream.config.upstream_ws_cdp_url
        return {"upstream": {"upstream_mode": "ws", "upstream_ws_cdp_url": launcher_local_loopback_cdp_url}} if launcher_local_loopback_cdp_url else {}

    def launch(self, config: LauncherConfig | dict[str, Any] | None = None) -> LaunchedBrowser:
        raise NotImplementedError(f"{type(self).__name__}.launch is not implemented.")

    def close(self) -> None:
        launched = self.launched
        self.launched = None
        if launched is not None:
            launched.close()

    def toJSON(self) -> dict[str, object]:
        return modCDPToJSON(
            self,
            {
                "state": {
                    "launched": self.launched is not None,
                    "cdp_url": self.launched.cdp_url if self.launched is not None else None,
                    "loopback_cdp_url": self.launched.loopback_cdp_url if self.launched is not None else None,
                    "cdp_listen_port": self.launched.cdp_listen_port if self.launched is not None else None,
                    "profile_dir": self.launched.profile_dir if self.launched is not None else None,
                    "browserbase_session_id": self.launched.browserbase_session_id if self.launched is not None else None,
                    "browserbase_session_url": self.launched.browserbase_session_url if self.launched is not None else None,
                    "browserbase_debug_url": self.launched.browserbase_debug_url if self.launched is not None else None,
                }
            },
        )


def merge_chrome_args(existing: list[str] | None = None, incoming: list[str] | None = None) -> list[str]:
    args = [*(existing or []), *(incoming or [])]
    load_extension_paths: list[str] = []
    merged: list[str] = []
    for arg in args:
        if not arg.startswith("--load-extension="):
            merged.append(arg)
            continue
        for extension_path in arg[len("--load-extension="):].split(","):
            if extension_path and extension_path not in load_extension_paths:
                load_extension_paths.append(extension_path)
    if load_extension_paths:
        first_url_index = next((index for index, arg in enumerate(merged) if not arg.startswith("-")), -1)
        load_extension_arg = f"--load-extension={','.join(load_extension_paths)}"
        if first_url_index == -1:
            merged.append(load_extension_arg)
        else:
            merged.insert(first_url_index, load_extension_arg)
    return merged


def _launcher_config(config: LauncherConfig | dict[str, Any] | None = None) -> LauncherConfig:
    if isinstance(config, LauncherConfig):
        return config
    return LauncherConfig.model_validate(config or {})


def resolveCdpWebSocketUrl(endpoint: str, name: str = "cdp_url") -> str:
    if endpoint.startswith(("ws://", "wss://")):
        return endpoint
    http_endpoint = endpoint if CDP_URL_SCHEME_RE.match(endpoint) else f"http://{endpoint}"
    with urllib.request.urlopen(f"{http_endpoint.rstrip('/')}/json/version", timeout=10) as response:
        version = json.loads(response.read().decode())
    cdp_url = version.get("webSocketDebuggerUrl") if isinstance(version, dict) else None
    if not isinstance(cdp_url, str) or not cdp_url:
        raise RuntimeError(f"{name} HTTP discovery returned no webSocketDebuggerUrl")
    return cdp_url

File: go/modcdp/launcher/BrowserLauncher.go
// MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
// - ./js/src/launcher/BrowserLauncher.ts
// - ./python/modcdp/launcher/BrowserLauncher.py
package launcher

import (
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"strings"
	"time"

	"github.com/browserbase/modcdp/go/modcdp/types"
)

type LauncherConfig = types.LauncherConfig
type InjectorConfig = types.InjectorConfig
type UpstreamTransportConfig = types.UpstreamTransportConfig

const DefaultChromeReadyTimeoutMS = 45_000
const DefaultChromeReadyPollIntervalMS = 100

var cdpHTTPClient = &http.Client{Timeout: 2 * time.Second}

func freePort() (int, error) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return 0, err
	}
	defer listener.Close()
	return listener.Addr().(*net.TCPAddr).Port, nil
}

func WebsocketURLFor(endpoint string) (string, error) {
	if strings.HasPrefix(endpoint, "ws://") || strings.HasPrefix(endpoint, "wss://") {
		return endpoint, nil
	}
	httpEndpoint := endpoint
	if !strings.Contains(endpoint, "://") {
		httpEndpoint = "http://" + endpoint
	}
	resp, err := cdpHTTPClient.Get(httpEndpoint + "/json/version")
	if err != nil {
		return "", fmt.Errorf("GET /json/version: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("GET %s/json/version -> %d", httpEndpoint, resp.StatusCode)
	}
	body, _ := io.ReadAll(resp.Body)
	var version map[string]any
	if err := json.Unmarshal(body, &version); err != nil {
		return "", fmt.Errorf("parse /json/version: %w", err)
	}
	wsURL, _ := version["webSocketDebuggerUrl"].(string)
	if wsURL == "" {
		return "", fmt.Errorf("cdp_url HTTP discovery returned no webSocketDebuggerUrl")
	}
	return wsURL, nil
}

type LaunchedBrowser struct {
	CDPURL                string `json:"cdp_url,omitempty"`
	CDPListenPort         int    `json:"cdp_listen_port,omitempty"`
	LoopbackCDPURL        string `json:"loopback_cdp_url,omitempty"`
	Close                 func() `json:"-"`
	ProfileDir            string `json:"profile_dir,omitempty"`
	BrowserbaseSessionID  string `json:"browserbase_session_id,omitempty"`
	BrowserbaseSessionURL string `json:"browserbase_session_url,omitempty"`
	BrowserbaseDebugURL   string `json:"browserbase_debug_url,omitempty"`
}

type BrowserLauncher struct {
	Config   LauncherConfig
	Launched *LaunchedBrowser
}

func NewBrowserLauncher(config LauncherConfig) BrowserLauncher {
	if config.LauncherBBBaseURL == "" {
		config.LauncherBBBaseURL = DefaultBrowserbaseLauncherBaseURL
	}
	return BrowserLauncher{Config: config}
}

func (l *BrowserLauncher) Update(config LauncherConfig) *BrowserLauncher {
	l.Config = mergeLaunchConfig(l.Config, config)
	return l
}

func (l BrowserLauncher) ConfigForUpstream() map[string]any {
	config := map[string]any{}
	upstreamWSCDPURL := firstString(launchedCDPURL(l.Launched), l.Config.LauncherRemoteCDPURL)
	if upstreamWSCDPURL != "" {
		config["upstream_ws_cdp_url"] = upstreamWSCDPURL
	}
	return config
}

func (l BrowserLauncher) ConfigForServer(upstreamConfig UpstreamTransportConfig) map[string]any {
	launcherLocalLoopbackCDPURL := ""
	if l.Launched != nil && l.Launched.LoopbackCDPURL != "" {
		launcherLocalLoopbackCDPURL = l.Launched.LoopbackCDPURL
	} else if upstreamConfig.UpstreamMode == "ws" && upstreamConfig.UpstreamWSCDPURL != "" {
		launcherLocalLoopbackCDPURL = upstreamConfig.UpstreamWSCDPURL
	}
	if launcherLocalLoopbackCDPURL != "" {
		return map[string]any{"upstream": map[string]any{"upstream_mode": "ws", "upstream_ws_cdp_url": launcherLocalLoopbackCDPURL}}
	}
	return map[string]any{}
}

func (l BrowserLauncher) Launch(config LauncherConfig) (*LaunchedBrowser, error) {
	return nil, fmt.Errorf("%T.Launch is not implemented", l)
}

func (l *BrowserLauncher) Close() {
	launched := l.Launched
	l.Launched = nil
	if launched != nil && launched.Close != nil {
		launched.Close()
	}
}

func (l BrowserLauncher) ToJSON() map[string]any {
	state := map[string]any{"launched": l.Launched != nil}
	if l.Launched != nil {
		state["cdp_url"] = l.Launched.CDPURL
		state["loopback_cdp_url"] = l.Launched.LoopbackCDPURL
		state["cdp_listen_port"] = l.Launched.CDPListenPort
		state["profile_dir"] = l.Launched.ProfileDir
		state["browserbase_session_id"] = l.Launched.BrowserbaseSessionID
		state["browserbase_session_url"] = l.Launched.BrowserbaseSessionURL
		state["browserbase_debug_url"] = l.Launched.BrowserbaseDebugURL
	}
	return types.ModCDPToJSON(l, types.ModCDPJSONConfig{State: state})
}

func mergeLaunchConfig(existing LauncherConfig, incoming LauncherConfig) LauncherConfig {
	merged := existing
	if incoming.LauncherLocalExecutablePath != "" {
		merged.LauncherLocalExecutablePath = incoming.LauncherLocalExecutablePath
	}
	if incoming.LauncherLocalCDPListenPort != 0 {
		merged.LauncherLocalCDPListenPort = incoming.LauncherLocalCDPListenPort
	}
	if incoming.LauncherLocalLoopbackCDP != nil {
		merged.LauncherLocalLoopbackCDP = incoming.LauncherLocalLoopbackCDP
	}
	if incoming.LauncherLocalUserDataDir != "" {
		merged.LauncherLocalUserDataDir = incoming.LauncherLocalUserDataDir
	}
	if incoming.LauncherLocalCleanupUserDataDir != nil {
		merged.LauncherLocalCleanupUserDataDir = incoming.LauncherLocalCleanupUserDataDir
	}
	if incoming.LauncherLocalChromeReadyTimeoutMS != 0 {
		merged.LauncherLocalChromeReadyTimeoutMS = incoming.LauncherLocalChromeReadyTimeoutMS
	}
	if incoming.LauncherLocalChromeReadyPollIntervalMS != 0 {
		merged.LauncherLocalChromeReadyPollIntervalMS = incoming.LauncherLocalChromeReadyPollIntervalMS
	}
	if incoming.LauncherLocalHeadless != nil {
		merged.LauncherLocalHeadless = incoming.LauncherLocalHeadless
	}
	if incoming.LauncherLocalSandbox != nil {
		merged.LauncherLocalSandbox = incoming.LauncherLocalSandbox
	}
	if len(incoming.LauncherLocalArgs) > 0 {
		merged.LauncherLocalArgs = mergeChromeArgs(existing.LauncherLocalArgs, incoming.LauncherLocalArgs)
	}
	if len(incoming.LauncherLocalExtraArgs) > 0 {
		merged.LauncherLocalExtraArgs = mergeChromeArgs(existing.LauncherLocalExtraArgs, incoming.LauncherLocalExtraArgs)
	}
	if incoming.LauncherRemoteCDPURL != "" {
		merged.LauncherRemoteCDPURL = incoming.LauncherRemoteCDPURL
	}
	if incoming.LauncherBBAPIKey != "" {
		merged.LauncherBBAPIKey = incoming.LauncherBBAPIKey
	}
	if incoming.LauncherBBBaseURL != "" {
		merged.LauncherBBBaseURL = incoming.LauncherBBBaseURL
	}
	if incoming.LauncherBBSessionID != "" {
		merged.LauncherBBSessionID = incoming.LauncherBBSessionID
	}
	if incoming.LauncherBBKeepAlive != nil {
		merged.LauncherBBKeepAlive = incoming.LauncherBBKeepAlive
	}
	if incoming.LauncherBBCloseSessionOnClose != nil {
		merged.LauncherBBCloseSessionOnClose = incoming.LauncherBBCloseSessionOnClose
	}
	if incoming.LauncherBBRegion != "" {
		merged.LauncherBBRegion = incoming.LauncherBBRegion
	}
	if incoming.LauncherBBTimeout != 0 {
		merged.LauncherBBTimeout = incoming.LauncherBBTimeout
	}
	if incoming.LauncherBBExtensionID != "" {
		merged.LauncherBBExtensionID = incoming.LauncherBBExtensionID
	}
	if incoming.LauncherBBBrowserSettings != nil {
		merged.LauncherBBBrowserSettings = incoming.LauncherBBBrowserSettings
	}
	if incoming.LauncherBBUserMetadata != nil {
		merged.LauncherBBUserMetadata = incoming.LauncherBBUserMetadata
	}
	if incoming.LauncherBBSessionCreateParams != nil {
		merged.LauncherBBSessionCreateParams = incoming.LauncherBBSessionCreateParams
	}
	return merged
}

func mergeChromeArgs(existing []string, incoming []string) []string {
	args := append(append([]string{}, existing...), incoming...)
	loadExtensionPaths := []string{}
	merged := []string{}
	for _, arg := range args {
		if !strings.HasPrefix(arg, "--load-extension=") {
			merged = append(merged, arg)
			continue
		}
		for _, extensionPath := range strings.Split(strings.TrimPrefix(arg, "--load-extension="), ",") {
			if extensionPath != "" && !containsString(loadExtensionPaths, extensionPath) {
				loadExtensionPaths = append(loadExtensionPaths, extensionPath)
			}
		}
	}
	if len(loadExtensionPaths) > 0 {
		loadExtensionArg := "--load-extension=" + strings.Join(loadExtensionPaths, ",")
		firstURLIndex := -1
		for index, arg := range merged {
			if !strings.HasPrefix(arg, "-") {
				firstURLIndex = index
				break
			}
		}
		if firstURLIndex == -1 {
			merged = append(merged, loadExtensionArg)
		} else {
			merged = append(merged[:firstURLIndex], append([]string{loadExtensionArg}, merged[firstURLIndex:]...)...)
		}
	}
	return merged
}

func containsString(values []string, needle string) bool {
	for _, value := range values {
		if value == needle {
			return true
		}
	}
	return false
}

func firstString(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}

func launchedCDPURL(launched *LaunchedBrowser) string {
	if launched == nil {
		return ""
	}
	return launched.CDPURL
}

func launchedProfileDir(launched *LaunchedBrowser) string {
	if launched == nil {
		return ""
	}
	return launched.ProfileDir
}
